File: nanopay/src/net/nanopay/blob/AmazonS3BlobStore.java
/**
 * @license
 * Copyright 2018 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

package net.nanopay.blob;

import com.amazonaws.auth.AWSStaticCredentialsProvider;
import com.amazonaws.auth.BasicAWSCredentials;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.amazonaws.services.s3.model.*;
import foam.blob.*;
import foam.core.X;
import org.apache.commons.io.IOUtils;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.util.Arrays;

public class AmazonS3BlobStore
  extends AbstractBlobService
{
  public static final int BUFFER_SIZE = 8192;

  protected String bucket_;
  protected AmazonS3 s3Client_ = null;

  public AmazonS3BlobStore(X x, String bucket, String accessKeyId, String secretKeyId) {
    this(x, "us-east-1", bucket, accessKeyId, secretKeyId);
  }

  public AmazonS3BlobStore(X x, String region, String bucket, String accessKeyId, String secretKeyId) {
    setX(x);
    bucket_ = bucket;
    BasicAWSCredentials credentials = new BasicAWSCredentials(accessKeyId, secretKeyId);
    s3Client_ = AmazonS3ClientBuilder.standard()
        .withCredentials(new AWSStaticCredentialsProvider(credentials))
        .withRegion(region)
        .build();
  }

  @Override
  public Blob put_(X x, Blob blob) {
    // check if instance of AmazonS3Blob
    if ( !(blob instanceof AmazonS3Blob) ) {
      throw new RuntimeException("Invalid blob");
    }

    // Check if delegate instance of FileBlob
    AmazonS3Blob s3Blob = (AmazonS3Blob) blob;
    if ( !(s3Blob.getDelegate() instanceof FileBlob) ) {
      throw new RuntimeException("Invalid blob");
    }

    // upload file to S3
    FileBlob fileBlob = (FileBlob) s3Blob.getDelegate();
    String key = s3Blob.getKey() + "/" + fileBlob.getFile().getName();
    PutObjectRequest request = new PutObjectRequest(bucket_, key, fileBlob.getFile())
        .withCannedAcl(CannedAccessControlList.PublicRead);
    s3Client_.putObject(request);

    // create identified blob with the key as the id
    IdentifiedBlob ret = new IdentifiedBlob();
    ret.setId(key);
    ret.setX(getX());
    return ret;
  }

  @Override
  public Blob find_(X x, Object id) {
    InputStream is = null;
    ByteArrayOutputStream os = null;

    try {
      String key = (String) id;
      GetObjectRequest request = new GetObjectRequest(bucket_, key);
      S3Object result = s3Client_.getObject(request);
      ObjectMetadata metadata = result.getObjectMetadata();

      is = result.getObjectContent();
      os = new ByteArrayOutputStream();

      int read = 0;
      int count = 0;
      long length = metadata.getContentLength();

      byte[] buffer = new byte[BUFFER_SIZE];
      while ( (read = is.read(buffer, 0, BUFFER_SIZE) ) != -1 && count <= length ) {
        os.write(buffer, 0, read);
        count += read;
      }

      return new ByteArrayBlob(os.toByteArray());
    } catch (Throwable t) {
      t.printStackTrace();
      throw new RuntimeException(t);
    } finally {
      IOUtils.closeQuietly(os);
      IOUtils.closeQuietly(is);
    }
  }

  @Override
  public String urlFor_(X x, Blob blob) {
    throw new UnsupportedOperationException("Unsupported operation: urlFor_");
  }
}

File: nanopay/src/net/nanopay/blob/AmazonS3BlobStore.js
/**
 * NANOPAY CONFIDENTIAL
 *
 * [2020] nanopay Corporation
 * All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of nanopay Corporation.
 * The intellectual and technical concepts contained
 * herein are proprietary to nanopay Corporation
 * and may be covered by Canadian and Foreign Patents, patents
 * in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from nanopay Corporation.
 */

foam.CLASS({
  package: 'net.nanopay.blob',
  name: 'AmazonS3BlobStore',
  extends: 'foam.blob.AbstractBlobService',

  javaImports: [
    'com.amazonaws.auth.AWSStaticCredentialsProvider',
    'com.amazonaws.auth.BasicAWSCredentials',
    'com.amazonaws.services.s3.AmazonS3ClientBuilder',
    'com.amazonaws.services.s3.model.*',
    'foam.blob.HashingOutputStream',
    'foam.blob.IdentifiedBlob',
    'foam.blob.InputStreamBlob',
    'foam.nanos.app.AppConfig',
    'org.apache.commons.io.IOUtils',
    'org.apache.commons.codec.binary.Hex',

    'java.io.File',
    'java.io.FileInputStream',
    'java.io.FileOutputStream',
    'java.io.InputStream'
  ],

  properties: [
    {
      class: 'String',
      name: 'region',
      documentation: 'S3 region'
    },
    {
      class: 'String',
      name: 'bucket',
      documentation: 'S3 bucket'
    },
    {
      class: 'String',
      name: 'accessKeyId',
      documentation: 'S3 access key id'
    },
    {
      class: 'String',
      name: 'secretKeyId',
      documentation: 'S3 secret key id'
    },
    {
      class: 'Object',
      name: 's3Client',
      documentation: 'Amazon S3 client',
      javaType: 'com.amazonaws.services.s3.AmazonS3',
      hidden: true,
      transient: true,
      javaFactory:
`BasicAWSCredentials credentials = new BasicAWSCredentials(getAccessKeyId(), getSecretKeyId());
return AmazonS3ClientBuilder.standard()
  .withCredentials(new AWSStaticCredentialsProvider(credentials))
  .withRegion(getRegion())
  .build();`
    }
  ],

  methods: [
    {
      name: 'put_',
      javaCode:
`if ( blob instanceof IdentifiedBlob ) {
  return blob;
}

// TODO: add a BlobInputStream as an adapter which
// implements the InputStream interface against the
// Blob interface. Then we can easily interface with
// any existing APIs that expect streams.

HashingOutputStream os = null;

try {
  File tmp = File.createTempFile("blob", ".tmp");
  os = new HashingOutputStream(new FileOutputStream(tmp));
  blob.read(os, 0, blob.getSize());
  os.close();

  // generate digest, create input stream, create metadata
  AppConfig config = (AppConfig) getX().get("appConfig");
  String key = config.getMode().name() + "/" + Hex.encodeHexString(os.digest());
  InputStream is = new FileInputStream(tmp);
  ObjectMetadata metadata = new ObjectMetadata();
  metadata.setContentLength(tmp.length());

  // send file
  PutObjectRequest request = new PutObjectRequest(getBucket(), key, is, metadata)
    .withCannedAcl(CannedAccessControlList.PublicRead);
  getS3Client().putObject(request);

  // create identified blob with the key as the id
  IdentifiedBlob ret = new IdentifiedBlob();
  ret.setId(key);
  ret.setX(getX());
  return ret;
} catch (Throwable t) {
  throw new RuntimeException(t);
} finally {
  IOUtils.closeQuietly(os);
}`
    },
    {
      name: 'find_',
      javaCode:
`String key = (String) id;
GetObjectRequest request = new GetObjectRequest(getBucket(), key);
S3Object result = getS3Client().getObject(request);
ObjectMetadata metadata = result.getObjectMetadata();
return new AmazonS3InputStreamBlob(result, metadata.getContentLength());`
    },
    {
      name: 'urlFor_',
      javaCode:
`if ( !(blob instanceof IdentifiedBlob) ) {
  return null;
}
return getS3Client().getUrl(getBucket(), ((IdentifiedBlob) blob).getId()).toString();`
    }
  ]
});
